flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -5554,40 +5542,55 @@ class ArrayExprLowering {
+      auto buffi = builder.create<fir::CoordinateOp>(loc, eleRefTy, buff,
+                                                     mlir::ValueRange{off});
+      fir::factory::genScalarAssignment(
+          builder, loc,
+          [&]() -> ExtValue {
+            if (len)
+              return fir::CharBoxValue(buffi, len);
+            return buffi;
+          }(),
+          v);
+      builder.create<fir::StoreOp>(loc, plusOne, buffPos);
+    };
+
     // Copy the value.
     exv.match(
-        [&](const mlir::Value &v) {
-          // Increment the buffer position.
-          auto plusOne = builder.create<mlir::arith::AddIOp>(loc, off, one);
-
-          // Grow the buffer as needed.
-          mem = growBuffer(mem, plusOne, limit, buffSize, eleSz);
-
-          // Store the element in the buffer.
-          mlir::Value buff =
-              builder.createConvert(loc, fir::HeapType::get(resTy), mem);
-          auto buffi = builder.create<fir::CoordinateOp>(loc, eleRefTy, buff,
-                                                         mlir::ValueRange{off});
-          fir::factory::genScalarAssignment(builder, loc, buffi, v);
-          builder.create<fir::StoreOp>(loc, plusOne, buffPos);
-        },
+        [&](mlir::Value) { doTrivialScalar(exv); },
         [&](const fir::CharBoxValue &v) {
-          // Increment the buffer position.
-          auto plusOne = builder.create<mlir::arith::AddIOp>(loc, off, one);
-
-          // Grow the buffer as needed.
-          mem = growBuffer(mem, plusOne, limit, buffSize, eleSz);
-
-          // Store the element in the buffer.
-          mlir::Value buff =
-              builder.createConvert(loc, fir::HeapType::get(resTy), mem);
-          mlir::Value buffi = computeCoordinate(buff, off);
-          llvm::SmallVector<mlir::Value> args = fir::runtime::createArguments(
-              builder, loc, memcpyType(), buffi, v.getAddr(), eleSz,
-              /*volatile=*/builder.createBool(loc, false));
-          createCallMemcpy(args);
-
-          builder.create<fir::StoreOp>(loc, plusOne, buffPos);
+          auto buffer = v.getBuffer();
+          if (fir::isa_char(buffer.getType())) {
+            doTrivialScalar(exv, eleSz);
+          } else {
+            // Increment the buffer position.
+            auto plusOne = builder.create<mlir::arith::AddIOp>(loc, off, one);
+
+            // Grow the buffer as needed.
+            mem = growBuffer(mem, plusOne, limit, buffSize, eleSz);
+
+            // Store the element in the buffer.
+            mlir::Value buff =
+                builder.createConvert(loc, fir::HeapType::get(resTy), mem);
+            mlir::Value buffi = computeCoordinate(buff, off);
+            llvm::SmallVector<mlir::Value> args = fir::runtime::createArguments(
+                builder, loc, memcpyType(), buffi, v.getAddr(), eleSz,
+                /*volatile=*/builder.createBool(loc, false));
+            createCallMemcpy(args);
```

Comment:
The code here looks strikingly similar to the code starting at line 5545.  Should it encapsulated?

Reply:
I noticed that as well. The prelude of the pattern being generated is the same (very similar), but they diverge at the end.